src/algebra/group_action_hom.lean

```diff
@@ -0,0 +1,259 @@
+structure mul_action_hom :=
+(to_fun : X → Y)
+(map_smul' : ∀ (m : M) (x : X), to_fun (m • x) = m • to_fun x)
+
+notation X ` →[`:25 M:25 `] `:0 Y:0 := mul_action_hom M X Y
+
+namespace mul_action_hom
+
+instance : has_coe_to_fun (X →[M] Y) :=
+⟨_, λ c, c.to_fun⟩
+
+variables {M X Y}
+
+@[simp] lemma map_smul (f : X →[M] Y) (m : M) (x : X) : f (m • x) = m • f x :=
+f.map_smul' m x
+
+@[ext] theorem ext : ∀ {f g : X →[M] Y}, (∀ x, f x = g x) → f = g
+| ⟨f, _⟩ ⟨g, _⟩ H := by { congr' 1, ext x, exact H x }
+
+theorem ext_iff {f g : X →[M] Y} : f = g ↔ ∀ x, f x = g x :=
+⟨λ H x, by rw H, ext⟩
+
+variables (M) {X}
+
+/-- The identity map is equivariant. -/
+protected def id : X →[M] X :=
+⟨id, λ _ _, rfl⟩
+
+@[simp] lemma id_apply (x : X) : mul_action_hom.id M x = x := rfl
+
+variables {M X Y Z}
+
+/-- Composition of two equivariant maps. -/
+def comp (g : Y →[M] Z) (f : X →[M] Y) : X →[M] Z :=
+⟨g ∘ f, λ m x, calc
+g (f (m • x)) = g (m • f x) : by rw f.map_smul
+          ... = m • g (f x) : g.map_smul _ _⟩
+
+@[simp] lemma comp_apply (g : Y →[M] Z) (f : X →[M] Y) (x : X) : g.comp f x = g (f x) := rfl
+
+@[simp] lemma id_comp (f : X →[M] Y) : (mul_action_hom.id M).comp f = f :=
+ext $ λ x, by rw [comp_apply, id_apply]
+
+@[simp] lemma comp_id (f : X →[M] Y) : f.comp (mul_action_hom.id M) = f :=
+ext $ λ x, by rw [comp_apply, id_apply]
+
+local attribute [instance] mul_action.regular
+
+variables {G} (H)
+
+/-- The canonical map to the left cosets. -/
+def to_quotient : G →[G] quotient_group.quotient H :=
+⟨coe, λ g x, rfl⟩
+
+@[simp] lemma to_quotient_apply (g : G) : to_quotient H g = g := rfl
+
+end mul_action_hom
+
+/-- Equivariant additive monoid homomorphisms. -/
+@[nolint has_inhabited_instance]
+structure distrib_mul_action_hom extends A →[M] B, A →+ B.
+
+/-- Reinterpret an equivariant additive monoid homomorphism as an additive monoid homomorphism. -/
+add_decl_doc distrib_mul_action_hom.to_add_monoid_hom
+
+/-- Reinterpret an equivariant additive monoid homomorphism as an equivariant function. -/
+add_decl_doc distrib_mul_action_hom.to_mul_action_hom
+
+notation A ` →+[`:25 M:25 `] `:0 B:0 := distrib_mul_action_hom M A B
+
+namespace distrib_mul_action_hom
+
+instance has_coe : has_coe (A →+[M] B) (A →+ B) :=
+⟨to_add_monoid_hom⟩
+
+instance has_coe' : has_coe (A →+[M] B) (A →[M] B) :=
+⟨to_mul_action_hom⟩
+
+instance : has_coe_to_fun (A →+[M] B) :=
+⟨_, λ c, c.to_fun⟩
+
+variables {M A B}
+
+@[simp] lemma coe_fn_coe (f : A →+[M] B) : ((f : A →+ B) : A → B) = f := rfl
+@[simp] lemma coe_fn_coe' (f : A →+[M] B) : ((f : A →[M] B) : A → B) = f := rfl
+
+@[ext] theorem ext : ∀ {f g : A →+[M] B}, (∀ x, f x = g x) → f = g
+| ⟨f, _, _, _⟩ ⟨g, _, _, _⟩ H := by { congr' 1, ext x, exact H x }
+
+theorem ext_iff {f g : A →+[M] B} : f = g ↔ ∀ x, f x = g x :=
+⟨λ H x, by rw H, ext⟩
+
+@[simp] lemma map_zero (f : A →+[M] B) : f 0 = 0 :=
+f.map_zero'
+
+@[simp] lemma map_add (f : A →+[M] B) (x y : A) : f (x + y) = f x + f y :=
```

Comment:
I'm always uneasy with such simp lemmas. It's not clear to me that the RHS is simpler.

Reply:
But it's what we do for other "algebraic" maps in the library, and it seems to work quite well. I think that 95% of the time I'm using this direction.